File: src/main/java/com/rtomyj/yugiohAPI/controller/BannedCardsController.java
package com.rtomyj.yugiohAPI.controller;

import com.rtomyj.yugiohAPI.dao.Dao.Status;
import com.rtomyj.yugiohAPI.helper.LogHelper;
import com.rtomyj.yugiohAPI.model.Card;
import com.rtomyj.yugiohAPI.service.BannedCardsService;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiResponse;
import io.swagger.annotations.ApiResponses;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import javax.servlet.http.HttpServletRequest;


@RequestMapping(path = "${ygo.endpoints.banned-cards-v1}", produces = "application/json; charset=utf-8")
@RestController
@CrossOrigin(origins = "*")
@Api(tags = "Ban List")
public class BannedCardsController {
	@Autowired
	private BannedCardsService bannedCardsService;

	@Autowired
	@Value("${ygo.endpoints.banned-cards-v1}")
	private String endPoint;

	@Autowired
	private HttpServletRequest request;

	private static final Logger LOG = LogManager.getLogger();

	private static final Map<String, Map<String, Map<String, List<Card>>>> BAN_LIST_CARDS_CACHE = new HashMap<>();



	/**
	 * @return item
	 */
	@ResponseBody
	@GetMapping(path = "{startDate}")
	@ApiOperation(value = "Retrieve the full ban list of a specified {startDate}", response = ResponseEntity.class, tags = "Ban List")
	@ApiResponses(value = {
		@ApiResponse(code = 200, message = "OK"),
		@ApiResponse(code = 204, message = "Request yielded no content"),
		@ApiResponse(code = 400, message = "Malformed request, make sure startDate is valid")
	})
	public ResponseEntity<Map<String, Map<String, List<Card>>>> getBannedCards(@PathVariable String startDate)
	{
		Pattern datePattern = Pattern.compile("[0-9]{4}-[0-9]{2}-[0-9]{2}");
		if (! datePattern.matcher(startDate).matches())
		{
			HttpStatus status = HttpStatus.BAD_REQUEST;
			LOG.info(LogHelper.requestInfo(request.getRemoteHost(), endPoint, String.format("Responding with: { %s }", status)));
			return new ResponseEntity<>(status);
		}


		if (BAN_LIST_CARDS_CACHE.get(startDate) != null)
		{
			HttpStatus status = HttpStatus.OK;
			LOG.info(LogHelper.requestInfo(request.getRemoteHost(), endPoint, String.format("Retrieved from BAN_LIST_CARDS_CACHE: Responding with: { %s }", status)));

			return new ResponseEntity<>(BAN_LIST_CARDS_CACHE.get(startDate), status);
		}
		else
		{
			Map<String, Map<String, List<Card>>> banList = new HashMap<>();
			Map<String, List<Card>> banListSections = new LinkedHashMap<>();

			banListSections.put("forbidden", bannedCardsService.getBanListByBanStatus(startDate, Status.FORBIDDEN));
			banListSections.put("limited", bannedCardsService.getBanListByBanStatus(startDate, Status.LIMITED));
			banListSections.put("semiLimited", bannedCardsService.getBanListByBanStatus(startDate, Status.SEMI_LIMITED));

			if (banListSections.get("forbidden").size() == 0 && banListSections.get("limited").size() == 0
					&& banListSections.get("semiLimited").size() == 0) {
				HttpStatus status = HttpStatus.NO_CONTENT;
				LOG.info(LogHelper.requestInfo(request.getRemoteHost(), endPoint, String.format("Responding with: { %s }", status)));
				return new ResponseEntity<>(status);
			} else {
				HttpStatus status = HttpStatus.OK;
				LOG.info(LogHelper.requestInfo(request.getRemoteHost(), endPoint, String.format("Responding with: { %s }", status)));
				banList.put("bannedCards", banListSections);

				BAN_LIST_CARDS_CACHE.put(startDate, banList);
				return new ResponseEntity<>(banList, status);
			}
		}
	}
}

File: src/main/kotlin/com/rtomyj/skc/find/BannedCardsController.kt
package com.rtomyj.skc.find

import com.rtomyj.skc.config.ReactiveMDC
import com.rtomyj.skc.config.SwaggerConfig
import com.rtomyj.skc.exception.SKCException
import com.rtomyj.skc.model.BanListInstance
import com.rtomyj.skc.util.constant.SKCRegex
import com.rtomyj.skc.util.constant.SwaggerConstants
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.constraints.Pattern
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

/**
 * Configures endpoint(s) that can be used to obtain information about cards for a particular ban list.
 */
@RestController
@RequestMapping(path = ["/ban_list"], produces = ["application/json; charset=UTF-8"])
@Validated
@Tag(name = SwaggerConstants.BAN_LIST_TAG_NAME)
class BannedCardsController
/**
 * Create object instance.
 * @param bannedCardsService Service object to use to accomplish functionality needed by this endpoint.
 */ @Autowired constructor(
  /**
   * Service object used to get information about banned cards from the database.
   */
  val bannedCardsService: BannedCardsService
) {

  companion object {
    private val log = LoggerFactory.getLogger(this::class.java.name)
  }

  /**
   * User can get the contents of a ban list by the start date of the ban list.
   * User data will be validated. A regular expression is used to ensure that the
   * user passed a valid date. The ban list cache will be utilized to speed up
   * delivery. If desired ban list contents are not in query , the ban list
   * contents will be fetched from DB.
   *
   * @param banListStartDate The date the desired ban list took effect.
   * @return ban list for specified ban list start date.
   */
  @ResponseBody
  @GetMapping(path = ["{banListStartDate}/cards"])
  @Operation(
    summary = "Retrieves information about a ban list using a valid effective ban list start date (use /api/v1/ban/dates to see a valid list of start dates).",
    tags = [SwaggerConstants.BAN_LIST_TAG_NAME]
  )
  @ApiResponse(responseCode = "200", description = SwaggerConfig.HTTP_200_SWAGGER_MESSAGE)
  @ApiResponse(responseCode = "400", ref = "badRequest")
  @ApiResponse(responseCode = "404", ref = "notFound")
  @ApiResponse(responseCode = "500", ref = "internalServerError")
  @Throws(
    SKCException::class
  )
  fun getBannedCards(
    @Parameter(
      description = SwaggerConstants.BAN_LIST_START_DATE_DESCRIPTION,
      example = "2020-04-01",
      required = true,
      schema = Schema(implementation = String::class)
    ) @Pattern(
      regexp = SKCRegex.DB_DATE, message = "Date doesn't have correct format."
    ) @PathVariable banListStartDate: String, @Parameter(
      description = SwaggerConstants.SAVE_BANDWIDTH_DESCRIPTION,
      required = false,
      schema = Schema(implementation = Boolean::class)
    ) @RequestParam(
      name = "saveBandwidth", required = false, defaultValue = "true"
    ) saveBandwidth: Boolean = true, @RequestParam(
      name = "format", required = true, defaultValue = "TCG"
    ) format: String = "TCG", @Parameter(
      description = SwaggerConstants.BAN_LIST_FETCH_ALL_DESCRIPTION,
      required = false,
      schema = Schema(implementation = Boolean::class)
    ) @RequestParam(
      name = "allInfo", required = false, defaultValue = "false"
    ) fetchAllInfo: Boolean = false
  ): Mono<ResponseEntity<BanListInstance>> = ReactiveMDC.deferMDC(Mono
      .fromCallable { bannedCardsService.getBanListByDate(banListStartDate, saveBandwidth, format, fetchAllInfo) }
      .map { banListInstance ->
        log.info(
          "Successfully retrieved ban list contents for ban list w/ start date {} with saveBandwidth as {} for format {}.",
          banListStartDate,
          saveBandwidth,
          format
        )
        ResponseEntity.ok(banListInstance)
      })
}